README.md
# Projet_Python
Projet python

# page_product.py : 


    >> product_details(id_product : str) <<

This function recover almost every product details in the json file. It takes one parameter : 

    - id_product  (string) : an unique value for retrieving the product in the json file.

This function returns the product name, the product price, the product description, the product stock, the product category, the list of best product.

    >> add_to_cart(id_user : str, id_product : str) :

This function add a product to the cart. It takes two parameters :

    - id_user (string) : an unique value to check if the cart exists.
    - id_product (string) : an unique value of the selected value to add in the cart.

This function returns 202 if the product is added to the cart, else it returns 404.

    >> best_product(best_product : bool, db_json) <<

This function recover every product that are best product and put them in a list. It takes two parameters : 

    - best_product (boolean) : this value is true if the product is in best_product else this value is not in best_product.
    - db_json (json) : this is the json you want to search for a best product.

This function returns the list of best products if the list exists, else it returns "".

# user.py : 


    >> register(id_user : str, password : str) <<

This function create a user in the json file. It takes two parameters : 

    - id_user (string) : an unique value for retrieving the user in the json file and also the name of the user.
    - password (string) : the password of the user.

This function returns "User already taken" if the id_user already exists in the json file, else it returns the function add_user_to_db() for creating the account.

    >> change_password(id_user : str, password : str) <<

This function update the json file by modifying the password of the selected user to the selected password. It takes two parameters :

    - id_user (string) : an unique value for retrieving the user in the json file and also the name of the selected user.
    - password (string) : the new password for the user.

This function returns "The selected user doesn't exist" if the id_user doesn't exist in the json file, else it returns informations on the selected users after updating the password.

    >> login(id_user : str, password : str) <<

This function check if the user exist in the json file. It takes two parameters :

    - id_user (string) : the selected user id to compare with the user id in the json file.
    - password (string) : the selected password to compare with the password of the user in the json file.

This function returns "You can enter" if the user id and the password exist in the json file, else it returns "Wrong username or password".

    >> add_user_to_db(file_json : str, id : str , password : str) <<

This function add a user in a json file. It takes three parameters :

    - file_json (string) : the json file where you want to add the user.
    - id (string) : the user id to add in the json file.
    - password (string) : the password to add in the json file.

This function returns the string : str(id) + " ok". It shows if the creation of the user is successful or not.

    >> check_id_user(id_user : str) <<

This function check if the id_user already exists in the json file. It takes one parameter : 

    - id_user (string) : the user you want to check with the json file.

This function returns the user id of the json file if it already exists, else it returns the string "You can create your account".

# list_products.py : 


    >> read_products(category: str): <<
 
This function create a user in the json file. It takes one parameter : 

    - category (string) : an value for retrieving the category in the json file.

This function Filter the products with category of the json file if it already exists, else it returns the string "The selected category doesn't exist".

     >> read_list_products(): <<

This function returns the products of the json file if it already exists, else it returns the string "The selected category doesn't exist".



HISTORY.md
## Add your updates here

## 0.1.0 (2022-09-26)

* First release

cart.py
from posixpath import join
from fastapi import FastAPI
import json

app = FastAPI()
# db_file_* is used to set the same db for each function (easy to change)
db_file_users = "users.json"
db_file_commands = "commands.json"
db_file_carts = "carts.json"
db_file_products = "products.json"

# Check if the cart of the user exist and if yes, it will return the associated cart
@app.get("/carts/{id_user}")
async def display_cart(id_user : str):
    f = open(db_file_carts, 'r')
    files = f.read()
    db_json = json.loads(files)
    for cart in db_json["carts"]:
        if cart["id_user"] == id_user:
            return cart["cart_content"]
    return "There is no cart for this user"

# Check if the cart of the user exist and if yes, it will check if the element exist in the cart and if yes, 1 element will be removed
@app.delete("/carts/{id_user}/delete/{id_product}")
async def delete_product(id_user : str, id_product : str):
    f = open(db_file_carts, 'r+')
    files = f.read()
    db_json = json.loads(files)
    f.close()
    for i in range(0, len(db_json["carts"])):
        if db_json["carts"][i]["id_user"] == id_user:
            for j in range(0, len(db_json["carts"][i])):
                if db_json["carts"][i]["cart_content"][j]["id_products"] == id_product:
                    if db_json["carts"][i]["cart_content"][j]["quantity"] <= 1:
                        db_json["carts"][i]["cart_content"].pop(j) #"product " + id_product + " deleted"
                    else:
                        db_json["carts"][i]["cart_content"][j]["quantity"] = int(db_json["carts"][i]["cart_content"][j]["quantity"]) -1
                    g = open(str(db_file_carts), 'w')
                    json.dump(db_json, g, sort_keys=True, indent = 4)
                    f.close()
                    return 200
    return 404

user.py
from posixpath import join
from fastapi import FastAPI
import json
 

app = FastAPI()
# db_file_* is used to set the same db for each function (easy to change)
db_file_users = "users.json"
db_file_commands = "commands.json"
db_file_carts = "carts.json"
db_file_products = "products.json"

# Create the user
@app.post("/users/{id_user}/{password}")
async def register(id_user : str, password : str):

        if id_user == check_id_user(id_user):
            return "User already taken"
        else:
            return add_user_to_db(db_file_users, id_user, password)

# Change password of the specified user
@app.put("/users/{id_user}/{password}")
async def change_password(id_user : str, password : str):

    # Recover db.json
    f = open(db_file_users, 'r')
    db_json = json.load(f)
    f.close()
    f = open(db_file_users, 'w')
    for i in range(0,len(db_json["users"])):
        if db_json["users"][i]["id_user"] == id_user:
            db_json["users"][i]["password"] = password
            json.dump(db_json, f, sort_keys=True, indent = 4)
            f.close()
            return db_json["users"][i]
    f.close()
    return "The selected user doesn't exist"

@app.get("/users/{id_user}/{password}")
async def login(id_user : str, password : str):
    f = open(db_file_users, 'r')
    files = f.read()
    db_json = json.loads(files)
    for i in range(0,len(db_json["users"])):
        if db_json["users"][i]["id_user"] == id_user and db_json["users"][i]["password"] == password:
            return "You can enter"
    return "Wrong username or password"

# Add the user to the json file
def add_user_to_db(file_json : str, id : str , password : str):
    f = open(str(file_json), 'r+')
    db_json = json.load(f)
    db_json["users"].append({
            "id_user":id,
            "password":password
        })
    f.close()
    f = open(str(file_json), 'w')
    json.dump(db_json, f, sort_keys=True, indent = 4)
    f.close()
    return str(id) + " ok"

# Check the id_user for creating new account
def check_id_user(id_user : str):
    
    f = open(db_file_users, 'r')
    files = f.read()
    db_json = json.loads(files)
    for i in range(0,len(db_json["users"])):
        if db_json["users"][i]["id_user"] == id_user:
            return db_json["users"][i]["id_user"]
    return "You can create your account"


# Check if the user exist and if yes, it will return the associated cart
@app.get("/carts/{id_user}")
async def display_cart(id_user : str):
    f = open(db_file_carts, 'r')
    files = f.read()
    db_json = json.loads(files)
    for cart in db_json["carts"]:
        if cart["id_user"] == id_user:
            return cart["cart_content"]
    return "There is no cart for this user"


@app.delete("/carts/{id_user}/delete/{id_product}")
async def delete_product(id_user : str, id_product : str):
    f = open(db_file_carts, 'r+')
    files = f.read()
    db_json = json.loads(files)
    f.close()
    for i in range(0, len(db_json["carts"])):
        if db_json["carts"][i]["id_user"] == id_user:
            for j in range(0, len(db_json["carts"][i])):
                if db_json["carts"][i]["cart_content"][j]["id_products"] == id_product:
                    if db_json["carts"][i]["cart_content"][j]["quantity"] <= 1:
                        db_json["carts"][i]["cart_content"].pop(j) #"product " + id_product + " deleted"
                    else:
                        db_json["carts"][i]["cart_content"][j]["quantity"] = int(db_json["carts"][i]["cart_content"][j]["quantity"]) -1
                    g = open(str(db_file_carts), 'w')
                    json.dump(db_json, g, sort_keys=True, indent = 4)
                    f.close()
                    return 200
    return 404

list_products.py
from fastapi import FastAPI
import json

app = FastAPI()
# db_file_* is used to set the same db for each function (easy to change)
db_file_users = "users.json"
db_file_commands = "commands.json"
db_file_carts = "carts.json"
db_file_products = "products.json"

@app.get("/products/{category}")
async def read_products(category: str):
    # Recover db.json
    f = open(db_file_products, 'r')
    files = f.read()
    db_json = json.loads(files)
    list_product = []
    for i in range(0, len(db_json["products"])):
        if db_json["products"][i]["category"] == category:
            list_product.append(db_json["products"][i])
    if list_product != []:
        return list_product
    return "The selected category doesn't exist"

@app.get("/products")
async def read_list_products(limit: int):
    f = open(db_file_products, 'r')
    files = f.read()
    db_json_out = {}
    db_json = json.loads(files)
    for i in range(1, limit +1):
        db_json_out[i] = db_json["products"][i-1]
    return db_json_out

# @app.patch("/commands/validation")
# async def valid_command(cart : str, commands str):
#     f = open(db_file_products, 'r')
#     files = f.read()
#     db_json_out = {}


page_product.py
from fastapi import FastAPI
import json

app = FastAPI()
# db_file_* is used to set the same db for each function (easy to change)
db_file_users = "users.json"
db_file_commands = "commands.json"
db_file_carts = "carts.json"
db_file_products = "products.json"

# Get the product name, price and description
@app.get("/product/{id_product}")
async def product_details(id_product : str):

    list_best_product = []
    product = ""

     # Recover db.json
    f = open(db_file_products, 'r')
    db_json = json.load(f)

    for i in range(0,len(db_json["products"])):
        list_best_product.append(best_product(db_json["products"][i]["best_product"], db_json["products"][i]))
        if db_json["products"][i]["id_product"] == id_product:
            product = db_json["products"][i]["product"]
            price = db_json["products"][i]["price"]
            description = db_json["products"][i]["description"]
            stock = db_json["products"][i]["stock"]
            category = db_json["products"][i]["category"]
    if product != "":
        return product, price, description, stock, category, list_best_product
    return "The selected product doesn't exist"

@app.post("/cart/{id_user}/product/{id_product}")
async def add_to_cart(id_user : str, id_product : str):
    product_found = ""
    f = open(db_file_carts, 'r')
    f_file = f.read()
    db_json = json.loads(f_file)
    f.close()
    g = open(db_file_products, 'r+')
    g_file = g.read()
    db_json2 = json.loads(g_file)
    g.close()
    for i in range(0,len(db_json["carts"])):
        if db_json["carts"][i]["id_user"] == id_user: # if the user already have a cart
            pass
        else : # if the user doesn't already have a cart
        # check for each products in the cart if he already exist
        for k in range(0, len(db_json["carts"][i]["cart_content"])): 
            if str(id_product) == db_json["carts"][i]["cart_content"][k]["id_products"]:
                product_found = "cart"
        # check for each products if he exist in products database
        for j in range(0, len(db_json2["products"])):
            if product_found: break    
            if (str(id_product) == db_json2["products"][j]["id_product"]):
                product_found = "product"
        # if the product exist in the cart:
        if product_found == "cart": 
            for j in range(0, len(db_json["carts"][i])):
                if db_json["carts"][i]["cart_content"][j]["id_products"] == id_product:
                    db_json["carts"][i]["cart_content"][j]["quantity"] += 1
                    break
        # if the product exist in the products database (but not in the cart)
        elif product_found == "product":
            for j in range(0, len(db_json["carts"][i])):
                if db_json["carts"][i]["cart_content"][j]["id_products"] == id_product:
                    db_json["carts"][i]["cart_content"][j]["quantity"] = 1
                    break
        else:
            return 404
        
    h = open(str(db_file_carts), 'w')
    json.dump(db_json, h, sort_keys=True, indent = 4)
    h.close()
    return 200
        

    # Recover the right cart for 
    # if id_user = user :
    #   Recover the right product for 
    #       If stock > 0 & cart[element] number < stock
    #           If quantity > 0 
    #               add 1 to the quantity in the cart (of the selected product <- user)
    #           else
    #           create a new product in cart_content
    #       Else
    #           return 404

    

def best_product(best_product : bool, db_json):

    if best_product == True:
        return db_json
    else:
        return 404

